### ISSUE TITLE
Use Windows 7 taskbar API to show progress

### ISSUE THREAD

**Issue by KOLANICH** [tqdm/tqdm#442]
Here is the manual how to use MS COM for accessing taskbar extensions and a precompiled ```type library``` for that.
https://www.neowin.net/forum/topic/716968-using-apis-in-a-objectaction-format/?tab=comments#comment-590434472

**Comment by vuki** [comment]
Proof of concept (Windows console):

```
import comtypes.client as cc
import win32api
import win32gui
import time
import tqdm

cc.GetModule('tl.tlb')
import comtypes.gen.TaskbarLib as tbl
taskbar = cc.CreateObject('{56FDF344-FD6D-11d0-958A-006097C9A090}', interface=tbl.ITaskbarList3)
taskbar.HrInit()

# find hWnd of the console
title = win32api.GetConsoleTitle()
tag = title + '___'
win32api.SetConsoleTitle(tag)
time.sleep(0.05)
hWnd = win32gui.FindWindow(None, tag)
win32api.SetConsoleTitle(title)
assert hWnd

# demo loop
for t in tqdm.trange(50):
    time.sleep(0.1)
    # this should be inside tqdm module:
    taskbar.SetProgressValue(hWnd, t, 50)
    taskbar.SetProgressState(hWnd, 0x2)
```

Adapted from:
https://stackoverflow.com/questions/1736394/using-windows-7-taskbar-features-in-pyqt
https://pastebin.com/f7ed294a8
https://support.microsoft.com/en-us/help/124103/how-to-obtain-a-console-window-handle-hwnd

**Comment by KOLANICH** [comment]
Error checking is also needed: Windows XP doesn't have this API.

**Comment by hildogjr** [comment]
This enhancement depends so much of the OS used. Ubuntu (Unity) also have some process that can be show on the program icon on the task bar.

**Comment by KOLANICH** [comment]
>Ubuntu (Unity) also have some process that can be show on the program icon on the task bar.

I guess it and everything else dbus-related deserves a separate issue.

**Comment by hildogjr** [comment]
The point here, mine, not that I want to create for the Unix too (now). But here have to be careful with this implementation, because the code will deal with different package/libraries in each OS.
And doing for Windows now, is interesting already leave in the tqdm package code something like:

    if Windows:
        your code propose
    else if Unix:
        # Comente: future implementation here

Just to keep the pythonization.

**Comment by KOLANICH** [comment]
Here is the patch.

<details>
<summary>the patch. Apply it</summary>
```diff
From ac1206e1e1d9121d3a642ca5b422e8182ae5c927 Mon Sep 17 00:00:00 2001
From: KOLANICH <KOLANICH@user.noreply.github.com>
Date: Fri, 12 Oct 2018 04:43:26 +0300
Subject: [PATCH] Implemented taskbar extensions.

---
 tqdm/_tqdm.py                                 |  11 ++
 .../PlatformSpecificProgressReporter.py       |  12 ++
 tqdm/platform_specific/__init__.py            |  37 +++++++
 .../windows_taskbar_extensions.py             | 103 ++++++++++++++++++
 4 files changed, 163 insertions(+)
 create mode 100644 tqdm/platform_specific/PlatformSpecificProgressReporter.py
 create mode 100644 tqdm/platform_specific/__init__.py
 create mode 100644 tqdm/platform_specific/windows_taskbar_extensions.py

diff --git a/tqdm/_tqdm.py b/tqdm/_tqdm.py
index 3df3621..a621ffb 100755
--- a/tqdm/_tqdm.py
+++ b/tqdm/_tqdm.py
@@ -109,6 +109,7 @@ class TqdmDefaultWriteLock(object):
         self.release()
 
 
+from .platform_specific import MetaReporter
 class tqdm(Comparable):
     """
     Decorate an iterable object, returning an iterator which acts exactly
@@ -871,6 +872,7 @@ class tqdm(Comparable):
         self.last_print_t = self._time()
         # NB: Avoid race conditions by setting start_t at the very end of init
         self.start_t = self.last_print_t
+        self.platform_specific_reporter=MetaReporter()
 
     def __len__(self):
         return self.total if self.iterable is None else \
@@ -959,6 +961,8 @@ Please use `tqdm_gui(...)` instead of `tqdm(..., gui=True)`
                             if self.pos:
                                 self.moveto(abs(self.pos))
                             # Print bar update
+                            
+                            self.platform_specific_reporter.progress(self.n, self.total)
                             sp(self.__repr__())
                             if self.pos:
                                 self.moveto(-abs(self.pos))
@@ -995,6 +999,7 @@ Please use `tqdm_gui(...)` instead of `tqdm(..., gui=True)`
             self.last_print_n = last_print_n
             self.n = n
             self.miniters = miniters
+            self.platform_specific_reporter.success()
             self.close()
 
     def update(self, n=1):
@@ -1049,6 +1054,7 @@ Please use `tqdm_gui(...)` instead of `tqdm(..., gui=True)`
                         self.moveto(abs(self.pos))
 
                     # Print bar update
+                    self.platform_specific_reporter.progress(self.n, self.total)
                     self.sp(self.__repr__())
 
                     if self.pos:
@@ -1090,6 +1096,7 @@ Please use `tqdm_gui(...)` instead of `tqdm(..., gui=True)`
         # Prevent multiple closures
         self.disable = True
 
+        self.platform_specific_reporter.clear()
         # decrement instance pos and remove from internal set
         pos = abs(self.pos)
         self._decr_instances(self)
@@ -1117,6 +1124,7 @@ Please use `tqdm_gui(...)` instead of `tqdm(..., gui=True)`
                 if self.last_print_n < self.n:
                     # stats for overall rate (no weighted average)
                     self.avg_time = None
+                    self.platform_specific_reporter.clear()
                     self.sp(self.__repr__())
                 if pos:
                     self.moveto(-pos)
@@ -1213,6 +1221,7 @@ Please use `tqdm_gui(...)` instead of `tqdm(..., gui=True)`
 
         if not nolock:
             self._lock.acquire()
+        self.platform_specific_reporter.clear()
         self.moveto(abs(self.pos))
         self.sp('')
         self.fp.write('\r')  # place cursor back at the beginning of line
@@ -1229,6 +1238,8 @@ Please use `tqdm_gui(...)` instead of `tqdm(..., gui=True)`
 
         if not nolock:
             self._lock.acquire()
+        
+        self.platform_specific_reporter.progress(self.n, self.total)
         self.moveto(abs(self.pos))
         self.sp(self.__repr__())
         self.moveto(-abs(self.pos))
diff --git a/tqdm/platform_specific/PlatformSpecificProgressReporter.py b/tqdm/platform_specific/PlatformSpecificProgressReporter.py
new file mode 100644
index 0000000..608654a
--- /dev/null
+++ b/tqdm/platform_specific/PlatformSpecificProgressReporter.py
@@ -0,0 +1,12 @@
+class PlatformSpecificProgressReporter:
+    def __init__(self):
+        raise NotImplementedError()
+    
+    def progress(self, current:int, total:int):
+        raise NotImplementedError()
+    
+    def fail(self):
+        raise NotImplementedError()
+    
+    def clear(self):
+        raise NotImplementedError()
\ No newline at end of file
diff --git a/tqdm/platform_specific/__init__.py b/tqdm/platform_specific/__init__.py
new file mode 100644
index 0000000..ef4f274
--- /dev/null
+++ b/tqdm/platform_specific/__init__.py
@@ -0,0 +1,37 @@
+__all__=("MetaReporter",)
+import sys
+import importlib
+from .PlatformSpecificProgressReporter import PlatformSpecificProgressReporter
+
+systemToReporterMapping={
+    "win32": ("windows_taskbar_extensions", )
+}
+
+reporterClassesToUse=[]
+if sys.platform in systemToReporterMapping:
+    for modName in systemToReporterMapping[sys.platform]:
+        mod=importlib.import_module("."+modName, __package__)
+        cls=getattr(mod, mod.__all__[0])
+        reporterClassesToUse.append(cls)
+
+class MetaReporter(PlatformSpecificProgressReporter):
+    def __init__(self):
+        self.reporters=[]
+        for cls in reporterClassesToUse:
+            self.reporters.append(cls())
+    
+    def progress(self, current:int, total:int):
+        for rep in self.reporters:
+            rep.progress(current, total)
+    
+    def fail(self):
+        for rep in self.reporters:
+            rep.fail()
+    
+    def success(self):
+        for rep in self.reporters:
+            rep.success()
+    
+    def clear(self):
+        for rep in self.reporters:
+            rep.clear()
diff --git a/tqdm/platform_specific/windows_taskbar_extensions.py b/tqdm/platform_specific/windows_taskbar_extensions.py
new file mode 100644
index 0000000..1fb069e
--- /dev/null
+++ b/tqdm/platform_specific/windows_taskbar_extensions.py
@@ -0,0 +1,103 @@
+__all__=("WindowsTaskbarProgressReporter",)
+from ctypes import c_wchar_p, c_int, c_ulonglong, POINTER
+import comtypes.client
+from comtypes import GUID
+from comtypes import IUnknown
+from ctypes import HRESULT
+from comtypes import helpstring
+from comtypes import COMMETHOD
+from enum import IntFlag
+from win32console import GetConsoleWindow
+from .PlatformSpecificProgressReporter import PlatformSpecificProgressReporter
+import win32gui, win32con
+
+dummyMethod=COMMETHOD([], None, '__dummyMethodDontUse__')
+
+class Library(object):
+	name = 'TaskbarLib'
+	_reg_typelib_ = ('{683BF642-E9CA-4124-BE43-67065B2FA653}', 1, 0)
+
+class ITaskbarList(IUnknown):
+	_case_insensitive_ = True
+	_iid_ = GUID('{56FDF342-FD6D-11D0-958A-006097C9A090}')
+	_idlflags_ = []
+
+class ITaskbarList2(ITaskbarList):
+	_case_insensitive_ = True
+	_iid_ = GUID('{602D4995-B13A-429B-A66E-1935E44F4317}')
+	_idlflags_ = []
+
+class ITaskbarList3(ITaskbarList2):
+	_case_insensitive_ = True
+	_iid_ = GUID('{EA1AFB91-9E28-4B86-90E9-9E9F8A5EEFAF}')
+	_idlflags_ = []
+
+ITaskbarList._methods_ = [
+	dummyMethod,
+	dummyMethod,
+	dummyMethod,
+	dummyMethod,
+	dummyMethod,
+]
+
+ITaskbarList2._methods_ = [
+	dummyMethod,
+]
+
+class TBPFlag(IntFlag):
+	noProgress = 0
+	indeterminate = 1
+	normal = 2
+	error = 4
+	paused = 8
+
+
+ITaskbarList3._methods_ = [
+	COMMETHOD([], HRESULT, 'SetProgressValue', ( ['in'], c_int, 'hwnd' ), ( ['in'], c_ulonglong, 'ullCompleted' ), ( ['in'], c_ulonglong, 'ullTotal' )),
+	COMMETHOD([], HRESULT, 'SetProgressState', ( ['in'], c_int, 'hwnd' ), ( ['in'], c_int, 'tbpFlags' )),
+	dummyMethod,
+	dummyMethod,
+	dummyMethod,
+	dummyMethod,
+	dummyMethod,
+	dummyMethod,
+	dummyMethod,
+	COMMETHOD([], HRESULT, 'SetOverlayIcon', ( ['in'], c_int, 'hwnd' ), ( ['in'], POINTER(IUnknown), 'hIcon' ), ( ['in'], c_wchar_p, 'pszDescription' )),
+	COMMETHOD([], HRESULT, 'SetThumbnailTooltip', ( ['in'], c_int, 'hwnd' ), ( ['in'], c_wchar_p, 'pszTip' )),
+	dummyMethod,
+]
+
+class iconz:
+    success = win32gui.LoadIcon(None, win32con.IDI_INFORMATION)
+    fail = win32gui.LoadIcon(None, win32con.IDI_ERROR)
+
+class WindowsTaskbarProgressReporter(PlatformSpecificProgressReporter):
+    def __init__(self):
+        self.tbl=comtypes.client.CreateObject('{56FDF344-FD6D-11d0-958A-006097C9A090}', interface=ITaskbarList3)
+    
+    def progress(self, current:int, total:int):
+        hwnd=GetConsoleWindow()
+        if total:
+            self.tbl.SetProgressState(hwnd, TBPFlag.normal)
+            self.tbl.SetProgressValue(hwnd, current, total)
+            #self.tbl.SetThumbnailTooltip(hwnd, "tqdm: "+str(current)+"/"+str(total))
+        else:
+            self.tbl.SetProgressState(hwnd, TBPFlag.indeterminate)
+            #self.tbl.SetThumbnailTooltip(hwnd, "tqdm: "+str(current))
+    
+    def fail(self):
+        hwnd=GetConsoleWindow()
+        self.tbl.SetProgressState(hwnd, TBPFlag.error)
+        #self.tbl.SetOverlayIcon(hwnd, iconz.fail, "An error occured during tqdmed loop")
+    
+    def success(self):
+        hwnd=GetConsoleWindow()
+        #self.tbl.SetOverlayIcon(hwnd, iconz.success, "tqdm loop succesfully finished")
+        self.tbl.SetProgressState(hwnd, TBPFlag.noProgress)
+        pass
+    
+    def clear(self):
+        hwnd=GetConsoleWindow()
+        self.tbl.SetProgressState(hwnd, TBPFlag.noProgress)
+        #self.tbl.SetOverlayIcon(hwnd, None, None)
+    
-- 
2.19.1.windows.1
```
</details>

Not very finished, but better than nothing. I have also got rid of binary type library file.

**Comment by KOLANICH** [comment]
@casperdcl, are you going to merge it?

**Comment by casperdcl** [comment]
one day when I find enough time I hope :P

**Comment by KOLANICH** [comment]
The patch have been updated, added a context manager and a check for WinXP.

**Comment by casperdcl** [comment]
so does #734 close this PR?

**Comment by KOLANICH** [comment]
>so does #734 close this PR?

Yes, and GH [will automatically close this issue once the PR is merged](https://help.github.com/en/articles/closing-issues-using-keywords)

**Comment by casperdcl** [comment]
ah srry I thought this was a PR, not an issue. thx

**Comment by somePythonProgrammer** [comment]
On Python, I would really recommend using this library:
[LINK](https://github.com/somePythonProgrammer/TbProgressLib)

**Comment by KOLANICH** [comment]
> committed 4 days ago

When I and @vuki developed this feature a few years ago, that lib was not available.  And your lib is inferior to my one because it requires a tlb file and because our code has slightly more feature. Also it looks like you have very little experience in working with git, since you store a wheel within a repo.

You also  may be interested in https://github.com/KOLANICH-libs/pantarei.py.

**Comment by somePythonProgrammer** [comment]
I have been on git for like, 2 months now.
